src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorComponentSearchEngine.cs

```diff
@@ -0,0 +1,93 @@
+
+        public DefaultRazorComponentSearchEngine(ProjectSnapshotManager projectSnapshotManager)
+        {
+            _projectSnapshotManager = projectSnapshotManager ?? throw new ArgumentNullException(nameof(projectSnapshotManager));
+        }
+
+        /// <summary>Search for a component in a project based on its tag name and fully qualified name.</summary>
+        /// <remarks>
+        /// This method makes several assumptions about the nature of components. First, it assumes that a component
+        /// a given name `Name` will be located in a file `Name.razor`. Second, it assumes that the namespace the
+        /// component is present in has the same name as the assembly its corresponding tag helper is loaded from.
+        /// Implicitly, this method inherits any assumptions made by TrySplitNamespaceAndType.
+        /// </remarks>
+        /// <param name="tagHelper">A TagHelperDescriptor to find the corresponding Razor component for.</param>
+        /// <returns>The corresponding DocumentSnapshot if found, null otherwise.</returns>
+        public override async Task<DocumentSnapshot> TryLocateComponentAsync(TagHelperDescriptor tagHelper)
+        {
+            if (tagHelper is null)
+            {
+                return null;
+            }
+
+            DefaultRazorTagHelperBinderPhase.ComponentDirectiveVisitor.TrySplitNamespaceAndType(tagHelper.Name, out _, out var typeSpan);
+            var typeName = tagHelper.Name.Substring(typeSpan.Start, typeSpan.Length);
+
+            foreach (var project in _projectSnapshotManager.Projects)
+            {
+                if (!project.FilePath.EndsWith($"{tagHelper.AssemblyName}.csproj", FilePathComparison.Instance))
+                {
+                    continue;
+                }
+
+                foreach (var path in project.DocumentFilePaths)
+                {
+                    // Rule out if not Razor component with correct name
+                    if (!IsPathCandidateForComponent(path, typeName))
+                    {
+                        continue;
+                    }
+
+                    // Get document and code document
+                    var documentSnapshot = project.GetDocument(path);
+                    var razorCodeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+                    if (razorCodeDocument is null)
+                    {
+                        continue;
+                    }
+
+                    // Make sure we have the right namespace of the fully qualified name
+                    if (!ComponentNamespaceMatchesFullyQualifiedName(razorCodeDocument, tagHelper.Name))
+                    {
+                        continue;
+                    }
+                    return documentSnapshot;
+                }
+            }
+            return null;
+        }
+
+        public static bool IsPathCandidateForComponent(string path, string componentName)
```

Comment:
Sorry I didn't mention this before, you should instead be taking in the DocumentSnapshot here and checking its `Kind` property. In .NET things are pretty much fully customizable which means a user could technically configure their app to have .cshtml files as components.

Reply:
```
            if (documentSnapshot.FileKind != FileKinds.Component)
            {
                return false;
            }
            var fileName = Path.GetFileNameWithoutExtension(documentSnapshot.FilePath);
            return fileName.Equals(fileName, FilePathComparison.Instance);
```

Reply:
Is this what you're thinking? It is a safe assumption to just check that the file name without extension matches up with the tag name?

Reply:
Yup!